### PR TITLE
WIP Azure fencing agent management

### PR DESCRIPTION
**Pull request by centiqmurphy**
## Commitment to Test

- Testing takes ?
- Reviewing ? files

### CODE REVIEW

File: util/create_fencing_agent.sh [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+# import common functions that are reused across scripts
+source util/common_utils.sh
+
+# import functions to create Service Principle
+
+# source util/create_service_principal.sh
+
+# name of the script where the auth info will be saved
+readonly auth_script='${service_principal_name}.sh'
```
Comment (ben-moss) [inline util/create_fencing_agent.sh:23]: I would prefix the SP auth script file name to have an easily identifiable initial set of characters... maybe all of them start with `set-clustering-auth-` so they're easily identifiable in the dir?

File: .gitignore [hunk 2]
```diff
@@ -19,3 +19,4 @@
 
 # service principal auth secrets
 set-sp.sh
+set-clustering-auth.sh
```
Comment (ben-moss) [inline .gitignore:22]:
```suggetsion

# ignore any clustering auth scripts
set-clustering-auth*.sh
```

File: util/create_fencing_agent.sh [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+# import common functions that are reused across scripts
+source util/common_utils.sh
+
+# import functions to create Service Principle
+
+# source util/create_service_principal.sh
+
+# name of the script where the auth info will be saved
+readonly auth_script="set-clustering-auth.sh"
```
Comment (ben-moss) [inline util/create_fencing_agent.sh:23]: Maybe either change from readonly and set with SID in name as part of `create_service_principal_script` or else define with ${1-} type default in the readonly definition.

File: util/create_fencing_agent.sh [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+	# ensure newlines in variable values are preserved
+	# backup current value to restore to afterwards
+	local ifs_backup="${IFS}"
+	IFS=$'\n'
+
+	# create the service principal and capture the output
+	sp_details=$(az ad sp create-for-rbac --name "${service_principal_name}")
+	local sp_details=$(az ad sp create-for-rbac --name "${service_principal_name}")
+	local sp_creation_status=$?=$?
```
Comment (robfallows) [inline util/create_fencing_agent.sh:66]: This is the reason your script is erroring when run.

```suggestion
ocal sp_creation_status=$?
```

**Comment by robfallows** [comment]
Your Test Instructions and Acceptance Criteria could do with some tidying up to get the layout looking right.

Also, you seem to be using tabs as indentation, which is an anti-pattern. Reconfiguring vscode to use spaces and then do `ctl-shift-P` -> `Convert Indentation to Spaces` on your edited files to fix.

File: util/create_fencing_agent.sh [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+
+function check_command_line_arguments()
+{
+    local args_count="$#"
+
+    # Check there's just a single argument provided
+    if [[ ${args_count} -ne 1 ]]; then
+        error_and_exit "You must specify a single command line argument for the SAP SID"
```
Comment (robfallows) [inline util/create_fencing_agent.sh:44]:
```suggestion
        error_and_exit "You must specify a single command line argument for the SAP SID. For example: $0 HN1"
```

File: util/create_fencing_agent.sh [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    fi
+}
+
+
+function create_service_principal_script()
+{
+    local service_principal_name="fencing-agent-$1"
```
Comment (robfallows) [inline util/create_fencing_agent.sh:51]:
```suggestion
    local SID=$1
    local service_principal_name="fencing-agent-${SID}"
```

File: util/create_fencing_agent.sh [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+readonly auth_script="set-clustering-auth-${1-}.sh"
+
+# link for service principal help
+readonly sp_link='https://docs.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli'
+
+function main()
+{
+    check_command_line_arguments "$@"
+
+    local SID="$1"
```
Comment (robfallows) [inline util/create_fencing_agent.sh:32]: Consider enforcing the case here. Maybe ensure lower case?

**Comment by robfallows** [comment]
When running `./util/create_fencing_agent.sh B4Z`:

```text
ERROR: Authorization file already exists: set-clustering-auth-B4Z.sh. Please reuse, move, or remove it.
```

This is acceptable, given I ran an earlier version of the code.

**Comment by robfallows** [comment]
WRT `az ad sp delete --id <principalId>`, there is nothing in the output JSON with the name `principalId`.

**Comment by robfallows** [comment]
WRT `az role definition delete --name "Linux Fence Agent Role"`:

`There are existing role assignments referencing role 03e5b287-4437-4e4c-8c69-ba33864e6eed. The role assignments must be deleted before the role can be deleted.`

File: util/create_fencing_agent.sh [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    # Define subscription id in json role template
+    az_subscription_id=$(echo "${subscription_id}" | sed 's/.\(.*\)/\1/' | sed 's/\(.*\)./\1/')
+    sed -i -e "s/SUBSCRIPTION_ID/${az_subscription_id}/" util/fencing_agent_role.json
+
+    # check role definition exsits
+    local fencing_template='Linux Fence Agent Role'
+    local template_file='util/fencing_agent_role.json'
+    local sp_prefix='http://'
+    az role definition create --role-definition "${template_file}" 2> /dev/null
+    # assign role to fencing service principle
```
Comment (robfallows) [inline util/create_fencing_agent.sh:96]:
```suggestion
    # assign role to fencing service principal
```

**Comment by robfallows** [comment]
Re-running the ` ./util/create_fencing_agent.sh T0P` (:hand: I used `T0P` to avoid conflicts with previously defined `B4Z`):

```text
Creating Azure Service Principal: fencing-agent-T0P...
Changing "fencing-agent-T0P" to a valid URI of "http://fencing-agent-T0P", which is the required format used for service principal names
'http://fencing-agent-T0P' already exists.
ERROR: There was a problem creating the service principal. If you are logged into Azure CLI, then it could relate to lack of admin/owner permissions. See https://docs.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli for further details.
```

That's a slightly different error to the expected one (`ERROR: Authorization file already exists: set-clustering-auth.sh. Please reuse, move, or remove it.` is displayed.)

File: util/fencing_agent_role.json [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    "Actions": [
+      "Microsoft.Compute/*/read",
+      "Microsoft.Compute/virtualMachines/deallocate/action",
+      "Microsoft.Compute/virtualMachines/start/action",
+      "Microsoft.Compute/virtualMachines/powerOff/action"
+    ],
+    "NotActions": [
+    ],
+    "AssignableScopes": [
+      "/subscriptions/0cbd16ac-c0b4-4c4a-9bca-4ca95477f0a9"
```
Comment (peterchoo) [inline util/fencing_agent_role.json:15]: It looks like the subscription ID placeholder has been overwritten again.

File: util/create_fencing_agent.sh [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    local client_id
+    local client_secret
+    subscription_id=$(az account show --query 'id')
+    tenant_id=$(az account show --query 'tenantId')
+    client_id=$(echo "${sp_details}"  | grep appId | sed -e 's/.*appId.:.\(.*\),/\1/')
+    client_secret=$(echo "${sp_details}" | grep password | sed -e 's/.*password.:.\(.*\),/\1/')
+
+    # create new script for authorization
+    cat <<- EOF > util/${auth_script}
+    export SAP_HANA_FENCING_AGENT_SUBSCRIPTION_ID=${subscription_id}
```
Comment (peterchoo) [inline util/create_fencing_agent.sh:79]:
```suggestion
      export SAP_HANA_FENCING_AGENT_SUBSCRIPTION_ID=${subscription_id}
```

File: util/create_fencing_agent.sh (hunk 7 above)
Comment (ben-moss) [reply util/create_fencing_agent.sh]: Trying to meet this requirement has broken Mac compatibility:

```
osx-bmoss:sap-hana work$ util/create_fencing_agent.sh BM1
util/create_fencing_agent.sh: line 28: ${1^^}: bad substitution
```

Comment (ben-moss) [reply util/create_fencing_agent.sh]: Is uppercase SID forced elsewhere in this codebase or in HANA?

File: util/create_fencing_agent.sh [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+#######################################################################################################################################################
+
+# exit immediately if an unset variable is used
+set -o nounset
+
+# import common functions that are reused across scripts
+source util/common_utils.sh
+
+# name of the script where the auth info will be saved
+readonly auth_script="set-clustering-auth-${1-}.sh"
```
Comment (robfallows) [inline util/create_fencing_agent.sh:19]: What is your expectation when using `${1-}`?

Comment (peterchoo) [reply util/create_fencing_agent.sh:19]: From the code, it looks like this was to insert the SID into the read only variable.  The `check_command_line_arguments` function will exit if there is not 1 argument for the SID

Comment (robfallows) [reply util/create_fencing_agent.sh:19]: However, `${1}` will do that. The `-` suffix has special meaning, but given the usage here, is just confusing.

Comment (robfallows) [reply util/create_fencing_agent.sh:19]: From <https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html>:

> When not performing substring expansion, using the form described below (e.g., ‘:-’), Bash tests for a parameter that is unset or null. Omitting the colon results in a test only for a parameter that is unset. Put another way, if the colon is included, the operator tests for both parameter’s existence and that its value is not null; if the colon is omitted, the operator tests only for existence.

Which is almost impossible to parse (or maybe I'm just tired), but clearly `${var-}` has meaning to distinguish it from `${var:-}` and `${var}`. I'd prefer to see `${1}` used - it's clear as to the meaning.

Comment (ben-moss) [reply util/create_fencing_agent.sh]: Using `-` on my advice - to default the value to empty string if not set. Otherwise errors happen. Maybe a better way, but used this elsewhere in the util scripts @robfallows
